mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java

```diff
@@ -165,21 +181,51 @@ public ReIndexerPerformer(MailboxManager mailboxManager,
             });
     }
 
-    private Mono<Result> handleMessageReIndexing(MailboxSession mailboxSession, Mailbox mailbox, MessageUid uid, ReprocessingContext reprocessingContext) {
-        return fullyReadMessage(mailboxSession, mailbox, uid)
-            .flatMap(message -> messageSearchIndex.add(mailboxSession, mailbox, message))
-            .thenReturn(Result.COMPLETED)
+    private Mono<MailboxMessage> fullyReadMessage(MailboxSession mailboxSession, Mailbox mailbox, MessageUid mUid) {
+        return mailboxSessionMapperFactory.getMessageMapper(mailboxSession)
+            .findInMailboxReactive(mailbox, MessageRange.one(mUid), MessageMapper.FetchType.Full, SINGLE_MESSAGE)
+            .next();
+    }
+
+    private Mono<ReIndexingEntry> createReindexingEntryFromFailure(ReIndexingFailure failure) {
+        MailboxSession mailboxSession = mailboxManager.createSystemSession(RE_INDEXER_PERFORMER_USER);
+
+        return mailboxSessionMapperFactory.getMailboxMapper(mailboxSession)
+            .findMailboxByIdReactive(failure.getMailboxId())
+            .flatMap(mailbox -> fullyReadMessage(mailboxSession, mailbox, failure.getUid())
+                .map(message -> new ReIndexingEntry(mailbox, mailboxSession, message)));
+    }
+
+    private Flux<ReIndexingEntry> findMessagesInMailbox(Mailbox mailbox, MailboxSession mailboxSession) {
+        MessageMapper messageMapper = mailboxSessionMapperFactory.getMessageMapper(mailboxSession);
+
+        return messageSearchIndex.deleteAll(mailboxSession, mailbox.getMailboxId())
+            .thenMany(messageMapper.listAllMessageUids(mailbox))
+            .flatMap(uid -> messageMapper.findInMailboxReactive(mailbox, MessageRange.one(uid), MessageMapper.FetchType.Full, UNLIMITED))
+            .map(message -> new ReIndexingEntry(mailbox, mailboxSession, message));
+    }
+
+    private Mono<Task.Result> reIndexMessages(Flux<ReIndexingEntry> entries, RunningOptions runningOptions, ReprocessingContext reprocessingContext) {
+        return throttle(entries, Duration.ofSeconds(1), runningOptions.getMessagesPerSecond())
+            .flatMap(entry -> reIndexMessage(entry.getMailboxSession(), entry.getMailbox(), reprocessingContext, entry.getMessage()))
```

Comment:
```suggestion
            .flatMap(entry -> reIndexMessage(entry.getMailboxSession(), entry.getMailbox(), reprocessingContext, entry.getMessage()), runningOptions.getMessagesPerSecond())
```

Overwize if the indexing speed don't keep up with specified rate we can do an unbounded number of concurrent reindexing, potentially harming the entire server performance.

With the change I propose we ensure the pending operations never overpass the rate specified by the admin.